### PR TITLE
Redirect open PRs when the target branch is deleted through merging a PR

### LINKED ISSUE #27062: Changing the destination branch of a pull request after deletion
### Feature Description

`branch3 -> branch2 -> branch1 -> master`
Now, if we have a chain of pull requests, we cannot merge `branch1` first, because all other requests to `branch1` will be closed without the possibility of reopening.

In bitbucket, this was solved by the fact that when deleting a `branch1`, all open pull requests change the target to the branch into which `branch1` was merged, in our case, this is the master.

### Screenshots

Here is an example in the form of a mind map.
Schema before merge `branch1`:
Before merge
After merge and deleting `branch1`:
After merge

### PR DESCRIPTION
**Pull request by tulzke**
Added a change the destination branch for those pull requests whose destination branch was deleted as a result of merge.
PR's HEAD changes to HEAD of the merged PR.

Resolves #27062

### CODE REVIEW

**Comment by denyskon** [comment]
Looks like a duplicate of https://github.com/go-gitea/gitea/pull/18478 to me, but with a slightly simpler implementation

**Comment by tulzke** [comment]
@denyskon , thx for reply.
Yes, it looks like it.
But the PR you are referring to does not look active. He is almost two years old, and he is still not in master.
But me and our team really needs this feature.
What needs to be done so that we finally see this feature in the next minor version?

**Comment by denyskon** [comment]
This implementation looks clean. I will try to test it out during the next few days :)
I think however that it counts as a new feature, so according to our policy it would not be released before Gitea 1.22

File: routers/web/repo/pull.go [hunk 1]
```diff
@@ -1313,6 +1313,31 @@ func MergePullRequest(ctx *context.Context) {
...
+			return
+		}
+
+		for _, prToHead := range pullRequestsToHead {
+			if prToHead.BaseRepoID != pr.BaseRepoID {
+				continue
+			}
+
+			err = prToHead.LoadIssue(ctx)
+			if err != nil {
```
Comment (denyskon) [inline routers/web/repo/pull.go]:
```suggestion
			if err := prToHead.LoadIssue(ctx); err != nil {
```

File: routers/web/repo/pull.go [hunk 2]
```diff
@@ -1313,6 +1313,31 @@ func MergePullRequest(ctx *context.Context) {
...
+				ctx.ServerError(fmt.Sprintf("LoadIssueForPullRequest[%d]", prToHead.ID), err)
+				return
+			}
+			prToHead.Issue.Repo = ctx.Repo.Repository
+
+			err = pull_service.ChangeTargetBranch(ctx, prToHead, ctx.Doer, pr.BaseBranch)
+			if err != nil {
```
Comment (denyskon) [inline routers/web/repo/pull.go]:
```suggestion
			if err := pull_service.ChangeTargetBranch(ctx, prToHead, ctx.Doer, pr.BaseBranch); err != nil {
```

**Review by denyskon — APPROVED** [review]
According to my tests, this feature seems to work great, and the implementation is a lot simpler than in the linked PR :+1:

**Comment by tulzke** [comment]
Hello @denyskon. Thanks for the review.
I re-read my code and realized that assigning `ctx.Repo.Repository` to `prTohead.Issue.Repo` is not valid in all cases.
I added a condition to check that both issues have the same `RepoId`. If this is not the case, then `Repo` is loaded from the database.

I apologize for making changes after the review.

**Review by a1012112796 — APPROVED** [review]

**Review by denyskon — APPROVED** [review]

File: routers/web/repo/pull.go [hunk 3]
```diff
@@ -1313,6 +1313,37 @@ func MergePullRequest(ctx *context.Context) {
...
+					ctx.ServerError(fmt.Sprintf("LoadRepoForIssue[%d]", prToHead.IssueID), err)
+					return
+				}
+			}
+
+			if err := pull_service.ChangeTargetBranch(ctx, prToHead, ctx.Doer, pr.BaseBranch); err != nil {
+				ctx.ServerError(fmt.Sprintf("ChangeTargetBranch[%d]", prToHead.ID), err)
+				return
+			}
+		}
```
Comment (delvh) [inline routers/web/repo/pull.go]: Could we please extract this code into a new subfunction?
Something like `redirectOpenPulls` or so?

**Comment by delvh** [comment]
Also, I'm currently wondering how edge cases will behave.
Do we run into any sort of security issue with this implementation, where someone can suddenly see something they are not supposed to see?
So, assume the following example:
We have
- public repo `A`
- fork `B` of `A`
- private repo `C`

The owner of `A` has been given collaboration permissions on `C`.
`A` opens a PR targeting `C` and `checks delete branch on merge`.
Similarly, `B` opens a PR targeting the base branch of `A`.
`A`s PR is merged.
So, is `B`s PR now targeting `C`?
Is that a security issue or isn't it?

**Comment by denyskon** [comment]
@delvh According to 
grafik

A fork of a private repository remains private. So how should `A` in your case become public? :thinking:

**Comment by kvaster** [comment]
I think that this PR is almost the same as mine (https://github.com/go-gitea/gitea/pull/18478), but it covers even less when mine:

This PR covers only `web/repo/pull`, but it do nothing with api - `routers/api/v1/repo/pull` and it should, I think.

And also both PRs do not cover following case:

Repo `A`.
Repo `B` is a fork of repo `A`
`PR1` is a PR from `B` to `A`
`PR2` is a PR from `B` to `PR` branch in `B`

What should happen in case `PR1` is merged and deleted ?
1) `PR2` should be closed
2) `PR2` should be retarget to `PR1`'s branch, but in repo `B` ?
3) `PR2` should be retarget to  `PR`'s target - i.e. to repo `A` ? But in this case we should also MOVE somehove `PR2` to repo `A`...

**Comment by kvaster** [comment]
So, I think, that we should:
* cover all entry points for merging PR's
* think of exact behaviour when base PR is targeting another repo

**Comment by kvaster** [comment]
> @denyskon , thx for reply. Yes, it looks like it. But the PR you are referring to does not look active. He is almost two years old, and he is still not in master. But me and our team really needs this feature. What needs to be done so that we finally see this feature in the next minor version?

The PR was complete from my point of view. Probably nobody cared about it and I had no time/wish to push this topic. I was just updating PR from time to time. Also I had no time (no experiance in go) to write unit tests for it.

**Comment by denyskon** [comment]
@kvaster

To me, it doesn't matter which PR is merged in the end. As it already has two approvals, it would maybe happen faster here, but I'm also open to reviewing the other PR if needed.

I'd say we need the functionality for the API, which can easily be provided by the function that delvh suggested to add. As for the example you brought up, I don't think we should retarget these PRs as it would basically mean that PRs end up in a repo they maybe weren't intended to, without the author even being notified about it.

**Comment by kvaster** [comment]
@denyskon, to me it always metters if someone is spending his time to help.

I already wrote my concerns about current status of this PR:

1) Not all entry points for merging PR's are covered.
2) Behaviour is unpredictable in case base PR is targeting another repo.
3) `pull_service.ChangeTargetBranch` error is not checked and merge can fail even if it's not an error

**Comment by kvaster** [comment]
This is not a competition. I really want this feature to be merged aswell. I' ve reopened my PR in case somebody will want to make a review.

https://github.com/go-gitea/gitea/pull/28686

**Comment by tulzke** [comment]
> 2. Behaviour is unpredictable in case base PR is targeting another repo.
You are wrong, the behavior is quite predictable. If a redirect has not been made for the PR, then it will be closed as it is now. 

I don't think redirecting PR to other repositories is a good idea.

> 3. `pull_service.ChangeTargetBranch` error is not checked and merge can fail even if it's not an error

The error is being checked. Why did you decide that this was not the case?

**Comment by tulzke** [comment]
@delvh in your case PR will be closed. The current behavior remained in place. If the PR has not been redirected to another branch, it will be closed.

File: routers/web/repo/pull.go (hunk 3 above)
Comment (tulzke) [reply routers/web/repo/pull.go]: I think it's a good idea. This will allow you to use redirection in the API, as @kvaster noted.

**Comment by kvaster** [comment]
@tulzke, seems I've missed - you're really checking if base repo is the same. 

For error checking - `ChangeTargetBranch` emits also following errors which are not errors:

```
issues_model.IsErrIssueIsClosed(err)
models.IsErrPullRequestHasMerged(err)
issues_model.IsErrPullRequestAlreadyExists(err)
```

Also `deleteBranch` is called in `reouter/web/repo/pull` in two places:

```
MergePullRequest
CleanUpPullRequest
```

both should be covered

Also `routers/api/v1/repo/pull` have method `MergePullRequest` - this one also should be covered.

As I said before - I really don't know why we should go through this once again.

**Comment by tulzke** [comment]
In the end, I don't care which PR is merged. I just want to see this feature in the next minor release.

I took the logic out into a separate function, and also added a call to the api.
Of the three voiced errors, it makes sense to check only one of them: issues_model.IsErrPullRequestAlreadyExists(err).

It doesn't make sense to check these two errors, because we only work with open PRS:

- issues_model.IsErrIssueIsClosed(err)
- models.IsErrPullRequestHasMerged(err)

**Comment by kvaster** [comment]
@tulzke, my thoughts about `issues_model.IsErrIssueIsClosed(err)` and `models.IsErrPullRequestHasMerged(err)` were only about what should happen if someone will merge child PR concurrently.

Also what do you think about `routers/api/v1/repo/pull` and `CleanUpPullRequest` method ?

**Comment by tulzke** [comment]
@kvaster 
> routers/api/v1/repo/pull

Yes, i added redirect to `MergePullRequest`.

> CleanUpPullRequest

I think you're right.
I hadn't thought about this method before, because I always deleted the branch during the merge.
In this case, the implementation almost completely copies yours, except for the presence of a feature-flag.

@denyskon @delvh @a1012112796 Can we review @kvaster's proposed PR? It looks like everything is ready there.

**Comment by denyskon** [comment]
Okay, I think we can agree on closing this in favor of #28686....

**Comment by tulzke** [comment]
> Okay, I think we can agree on closing this in favor of #28686....

Good news, thanks

**Pull request closed without merging**